### DISCUSSION TITLE
What is the level of security of Prisma ?

### DISCUSSION THREAD

**Discussion by MarcoBaroniGitHub** [prisma/prisma#25443] [Q&A]
### Question

Hello,

We are considering moving forward with Prisma for our project, but before doing so, we would like to have a complete understanding of the security capabilities of the framework. Specifically, we would appreciate some insights on the following:

SQL Injection: How does Prisma handle SQL injection attacks? What built-in mechanisms are in place to prevent such vulnerabilities, and are there any specific best practices we should follow to ensure maximum security?

Data validation: Does Prisma provide automatic validation of data before it's inserted or queried from the database? Are there customizable validation mechanisms available?

Sensitive data: How does Prisma handle sensitive data (e.g., encryption, hashing)? What features are available for data security both at rest and in transit?

Access control: How can we ensure that data access is properly controlled and restricted based on user roles or permissions within Prisma?

Security audits: Are there any security audits or certifications that Prisma has undergone or guidelines that you recommend following?

Thank you for your help!

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hello @MarcoBaroniGitHub 👋 

We are glad to hear you are considering using Prisma for your Project.

To address your questions regarding Prisma's security capabilities:
> SQL Injection

Prisma provides several mechanisms to prevent SQL injection attacks. The primary method is through the use of prepared statements, which are automatically used when you perform queries using Prisma's ORM models. When raw SQL queries are necessary, Prisma offers methods like `$queryRaw` and `$executeRaw` that can be used safely with tagged templates or parameterized queries. These methods ensure that variables are properly escaped, mitigating the risk of SQL injection. However, using `$queryRawUnsafe` or `$executeRawUnsafe` can increase the risk of SQL injection if not handled carefully. For more details, you can refer to the [documentation on SQL injection prevention](https://www.prisma.io/docs/orm/prisma-client/using-raw-sql/raw-queries#sql-injection-prevention).

> Data Validation

Prisma does not automatically validate data before it is inserted or queried from the database. However, you can implement custom validation logic in your application code before interacting with the database. This allows you to tailor the validation to your specific requirements.

> Sensitive Data

Prisma does not provide built-in features for encrypting or hashing sensitive data. You would need to implement these security measures in your application logic. For data security in transit, you should ensure that your database connection is configured to use SSL/TLS.

> Access Control

Prisma itself does not handle access control based on user roles or permissions. You would need to implement access control logic in your application layer, ensuring that only authorized users can perform certain operations.

> Security Audits

Prisma has achieved several security certifications, including SOC2 Type II, HIPAA, GDPR, and ISO 27001:2022. These certifications demonstrate Prisma's commitment to providing secure and reliable software solutions. You can read more about these achievements in the [blog post on compliance](https://www.prisma.io/blog/compliance-reqs-complete).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
